### ISSUE TITLE
Partially update a field.

### ISSUE THREAD

**Issue by Rogerooo** [mbdavid/LiteDB#918]
I'm trying to edit a database that has a field to hold information for specific folder paths on my pc. However those folders were moved and now the paths are pointing to the wrong place. Is it possible to use db.collection.update to partially update the field entries or do i need to edit the db file another way? 

Something like:
`db.folders.update Directory = "D:\\MyFolder\\*" where "C:\\MyFolder\\*"`
In this case the "*" sign indicates various other subfolders inside MyFolder. I would like to know if there is a away to achieve something like this thru the shell. Thanks in advance and great work you have here!

**Comment by mbdavid** [comment]
Yes, you can use some BsonExpressions for that:

In shell, you can use this syntax (if you are only change from C to D):

`db.folders.update $.Directory = "D" + SUBSTRING($.Directory, 1) where SUBSTRING($.Directory, 1, 12) = "C:\\MyFolder\\"`

(ok, in this case will be much better use REPLACE function, but I think that I miss this string function - will add soon)

**Comment by Rogerooo** [comment]
Thanks for the info, missed the Expressions wiki page. However I don't know if my syntax is wrong but the shell gives me an Unexpected token error on the + sign. This is my input (different letters but same concept):

`db.games.update $.InstallDirectory = "I" + SUBSTRING($.Directory, 1) where SUBSTRING($.InstallDirectory, 1, 11) = "H:\\_Backup\\"`

I'm trying to manipulate my Playnite db after changing files directories and this is the quickest way to do it. Also would something like this work for same drive changes:

`db.games.update $.InstallDirectory = "H:\\AnotherFolder\\" + SUBSTRING($.Directory, 1, 17) where SUBSTRING($.InstallDirectory, 1, 11) = "H:\\_Backup\\"`

**Comment by Rogerooo** [comment]
Hum, it now gives me an Unexpected JSON token: 'H error. I also tried double quoting `"H:\\_Backup\\"` and the result was [1]: 0 so no entries were changed right? I think i've took too much of your time with my lack of programming knowledge and for now i'll just edit the paths one by one in Playnite. Thanks for the help though and keep up the good work ;)

**Comment by mbdavid** [comment]
Ok, now I run on my tests here (problem still single/double quote) - In v4, you must represent, in expression, string as `'` but in JSON string value must be in double `"`.


```
> db.col1.insert {a:"h:\\temp\\file.txt"}
[1]: {"$oid":"5a8cc40737ca92288816bbf5"}
> db.col1.find
[1]: {"_id":{"$oid":"5a8cc40737ca92288816bbf5"},"a":"h:\temp\file.txt"}
> db.col1.update $.a = 'I' + SUBSTRING($.a, 1) where SUBSTRING($.a, 0, 1) = "h"
[1]: 1
> db.col1.find
[1]: {"_id":{"$oid":"5a8cc40737ca92288816bbf5"},"a":"I:\temp\file.txt"}
```

**Comment by Rogerooo** [comment]
It must be something wrong i'm doing or the way the db works but double quoting the JSON string gives me the [1]:0 result and nothing is changed. Heres an example:
```
> db.games.find InstallDirectory = "H:\\_Backup\\_Jogos\\_Roms\\Nintendo Wii"
[1]: {
    "_id": 153,
    "_type": "Playnite.Models.Game, Playnite",
    "Description": "Heading back to the early days of the Dragon Ball story, players control Kid Goku as he sets out on fantastic adventures with his friends. The title faithfully reproduces the uniqueness and special humour of the early Dragon Ball series, while adding an original blend of action and adventure to the mix. With the original Japanese music and voiceover included for maximum authenticity, Dragon Ball: Revenge of King Piccolo offers fans a dazzling new experience with the characters and universe they know and love.",
    "Developers":
    [
        "Media.Vision"
    ],
    "Genres":
    [
        "Fighting",
        "Platform",
        "Hack and slash/Beat 'em up",
        "Adventure"
    ],
    "Hidden": false,
    "Favorite": false,
    "Icon": "images/custom/46d5d780-426e-4a01-89a2-ac89bf592515.png",
    "Image": "images/custom/rwaeyklle4svchq77xem.jpg",
    "InstallDirectory": "H:\_Backup\_Jogos\_Roms\Nintendo Wii",
    "IsoPath": "H:\_Backup\_Jogos\_Roms\Nintendo Wii\Dragon Ball - Revenge of King Piccolo.iso",
    "Name": "Dragon Ball: Revenge of King Piccolo",
    "ProviderId": "9dfeda46-f2f2-4268-a9ab-bf03d104ad7f",
    "PlayTask":
    {
        "Type": "Emulator",
        "IsPrimary": false,
        "OverrideDefaultArgs": false,
        "IsBuiltIn": false,
        "EmulatorId": {"$oid": "5a5191b8ac455714b0a49841"},
        "EmulatorProfileId": {"$oid": "5a5191b4ac455714b0a4983e"}
    },
    "Provider": "Custom",
    "PlatformId": {"$oid": "5a3fdccdac45573f7c56e8bd"},
    "Publishers":
    [
        "Namco Bandai Games",
        "Bandai"
    ],
    "ReleaseDate": {"$date": "2009-07-22T23:00:00.0000000Z"},
    "Tags":
    [
        "Single Player",
        "Multiplayer"
    ],
    "Links":
    [
        {
            "Name": "Wikipedia",
            "Url": "https://en.wikipedia.org/wiki/Dragon_Ball:_Revenge_of_King_Piccolo"
        },
        {
            "Name": "Wikia",
            "Url": "http://dragonball.wikia.com/wiki/Dragon_Ball:_Revenge_of_King_Piccolo"
        }
    ]
}
> db.games.update $InstallDirectory = 'I' + SUBSTRING($.InstallDirectory, 1) where SUBSTRING($.InstallDirectory, 1, 1) = "H"
[1]: 0
> db.games.update $InstallDirectory = 'I' + SUBSTRING($.InstallDirectory, 1) where SUBSTRING($.InstallDirectory, 1, 1) = "H:\\_Backup\\_Jogos\\_Roms\\Nintendo Wii"
[1]: 0
```
I'm inexperienced with this kind of stuff so i'll just leave the information here in case someone is trying the same thing and change the directories in Playnite myself, it's a bit more time consuming but hopefully i'll have the database ready before i die of boredom eheh. Thanks a lot for the help and your time man really appreciate it.

**Comment by mbdavid** [comment]
Are you missing `.` after `$`... and `REPLACE` syntax are: `REPLACE(str, index, [length])` and index starts in `0`. So, your update must do:

```
> db.games.update $.InstallDirectory = 'I' + SUBSTRING($.InstallDirectory, 1) where SUBSTRING($.InstallDirectory, 0, 1) = "H"

> db.games.update $InstallDirectory = 'I' + SUBSTRING($.InstallDirectory, 1) where SUBSTRING($.InstallDirectory, 1) = "H:\\_Backup\\_Jogos\\_Roms\\Nintendo Wii"
```

Se tu tens um diretório chamado "Jogos", fala portugues 😄 Abraço,

**Comment by Rogerooo** [comment]
*Huge facepalm* that sneaky `.` was the problem eheh. I just tried it as you suggested and it worked flawlessly...until I opened playnite and the changes break the database, I get a duplication error on the _id indexes, don't know how it happened as I just changed the InstallDirectory field but since I realized this is way over my head i'll just leave the programming stuff to the guys that really know what they are doing eheh. I've asked on reddit and Playnite's dev Crowcz is working on a solution for this kind of task so i think i'll just wait until he's done his magic :D

Ahah não tinha reparado que eras do Brasil, eu sou portuga eheh, escrevi em inglês para ser mais fácil para quem tiver a mesma dúvida mas é bom saber que há muito talento nesta área no país irmão. Parabéns pelo projecto tá muito bacana, bem organizado e bem documentado. Um abraço deste lado do oceano e desculpa outra vez estar a chatear com a minha burrice :D
